stm32 xspi bindings fixes dtc warnings

FRASTM opened the pull request:
This PR is for fixing the https://github.com/zephyrproject-rtos/zephyr/issues/88404  for the XSPI compatible

It changes the way the "st,stm32-xspi-nor" compatible declares the external flash base address and size.

The **"st,stm32-xspi"** node, gives the external memory base address and maximum allocated space (in bytes):
```
		xspi1: spi@47001400 {
			compatible = "st,stm32-xspi";
			reg = <0x47001400 0x400>, <0x90000000 0x10000000>;
```
The **"st,stm32-xspi-nor"** node, gives the external NOR device size in Bits through the `size` property
```
	mx25lm51245: ospi-nor-flash@0 {
		compatible = "st,stm32-xspi-nor";
		reg = <0>;
		size = <0x20000000>; /* 512 Mbits */
```

FRASTM committed:
dts: bindings: flash controller size of the stm32 xspi nor

This PR adds the size in Bits of the flash nor memory
for the st,stm32-xspi-nor compatible

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
dts: arm: stm32N6 reg definition for the st,stm32-xspi compatible

The st,stm32-xspi compatible is defining the reg property
with the register address and size at first index
followed by the external memory base address and max allocated
size. For the stm32N6 serie,
xspi1 is addressing max 256 MBytes from 0x90000000
xspi2 is addressing max 256 MBytes from 0x70000000

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
dts: arm: stm32H5 reg definition for the st,stm32-xspi compatible

The st,stm32-xspi compatible is defining the reg property
with the register address and size at first index
followed by the external memory base address and max allocated
xspi1 is addressing max 256 MBytes from 0x90000000

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: flash: stm32 xspi driver size and address of the external NOR

New property of the st,stm32-xspi-nor compatible gives
the external NOR flash in bits.
The property of the st,stm32-xspi compatible gives
the external NOR flash base address

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
boards: st: stm32N6 boards xspi-nor-flash DTS configuration

This PR defines the "st,stm32-xspi-nor" compatible Node
and the "st,stm32-xspi-psram" compatible Node
in conformance to the DTS specifications
Includes the size property (in Bits) of the external memory device

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
boards: st: stm32H573 disco kit xspi-nor-flash DTS configuration

This PR defines the "st,stm32-xspi-nor" compatible Node
in conformance to the DTS specifications
Includes the size property (in Bits) of the external NOR device

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
samples: code_relocation_nocopy: update macro for flash size and address

In case of the st,stm32-xspi-nor compatible
new property and node definitions will requires new macro
to get the external NOR flash base address and size

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
No warnig during west build  -b stm32h573i_dk samples/application_development/code_relocation_nocopy
```
*** Booting Zephyr OS build v4.1.0-2355-g18aeeafb96ea ***
Address of main function 0x80005e1
Address of function_in_ext_flash 0x90000001
Address of var_ext_sram_data 0x200000a0 (10)
Address of function_in_sram 0x20000001
Address of var_sram_data 0x200000a4 (10)
Hello World! stm32h573i_dk
```

erwango commented on boards/st/stm32h573i_dk/stm32h573i_dk.dts:
```diff
@@ -273,9 +273,10 @@
 		compatible = "st,stm32-xspi-nor";
-		reg = <0x90000000 DT_SIZE_M(64)>; /* 512 Mbits */
+		reg = <0>;
+		size = <0x20000000>; /* 512 Mbits */
```
Nit: use `DT_SIZE_K/M()` macros everywhere possible to ease reading.

FRASTM replied:
done

erwango commented on dts/arm/st/h5/stm32h562.dtsi:
```diff
@@ -260,7 +260,7 @@
 		xspi1: spi@47001400 {
 			compatible = "st,stm32-xspi";
-			reg = <0x47001400 0x400>;
+			reg = <0x47001400 0x400>, <0x90000000 0x10000000>;
```
Nit: use `DT_SIZE_K/M()` macros everywhere possible (at least for large numbers)

FRASTM replied:
done

erwango reviewed (changes requested):
Please update migration guide, as this change will break out of tree users.

FRASTM committed:
doc: release-notes-4-2: new DTS properties for the xSPI of STM32 devices

Change to apply on the DTS of STM32 soc and boards with xSPI nodes

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
Adding text for the doc: release-notes-4-2:

rruuaanng commented:
If possible, making multiple commits in this PR could help reduce email noise :)

erwango commented on boards/st/nucleo_n657x0_q/nucleo_n657x0_q_common.dtsi:
```diff
@@ -210,9 +210,10 @@
 		compatible = "st,stm32-xspi-nor";
-		reg = <0x70000000 DT_SIZE_M(64)>; /* 512 Mbits */
+		reg = <0>;
+		size = <DT_SIZE_M(512)>; /* 512 Mbits */
```
Is it ?
[Common serial flash devices binding jedec,jesd216.yaml](https://github.com/zephyrproject-rtos/zephyr/blame/main/dts/bindings/mtd/jedec%2Cjesd216.yaml#L25) defines it in bits as well:
```
  size:
    type: int
    description: flash capacity in bits
```

FRASTM commented on doc/releases/migration-guide-4.2.rst:
```diff
@@ -286,6 +286,16 @@ SPI
+* For STM32, the external memory base address and max possible size is given by the reg property
+  of the compatible = "st,stm32-xspi"; (usually in the device DTS).
+  The reg property of the compatible = "st,stm32-xspi-nor"; (usually in the board DTS)
+  is kept <0> and the <size> property gives the actual size of the memory device.
```
changed

FRASTM commented on doc/releases/migration-guide-4.2.rst:
```diff
@@ -286,6 +286,16 @@ SPI
+xSPI
+===
+
+* For STM32, the external memory base address and max possible size is given by the reg property
```
changed

erwango commented on doc/releases/migration-guide-4.2.rst:
```diff
@@ -306,6 +306,16 @@ SPI
+xSPI
+===
+
+* For STM32, the external memory base address and max possible size was given by the reg property
+  of the compatible = "st,stm32-xspi"; (usually in the device DTS).
+  The reg property of the compatible = "st,stm32-xspi-nor"; (usually in the board DTS)
+  is now given <0> and the <size> property gives the actual size of the memory device in bits.
```
Let's be a bit more specific:

* On STM32 devices, external memories device tree descriptions for size and address are now split in two separate
  properties to comply with specification recommendations.
  For instance, following external flash description
      reg = <0x70000000 DT_SIZE_M(64)>; /* 512 Mbits */
  is changed to
      reg = <0>;
		size = <DT_SIZE_M(512)>; /* 512 Mbits */
 Note that the <size> property gives the actual size of the memory device in bits.
 Previous mapping address information is now described in xspi node at SoC dtsi level.

FRASTM committed:
drivers: memc: stm32 xspi driver size and address of the external PSRAM

New property of the st,stm32-xspi-psram compatible gives
the external PSRAM memory in bits.
The property of the st,stm32-xspi compatible gives
the external PSRAM memory base address

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
dts: bindings: memory controller size of the stm32 xspi psram

This PR adds the size in Bits of the PSRAM  memory
for the st,stm32-xspi-psram compatible

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
ci failure : https://github.com/zephyrproject-rtos/zephyr/issues/89081
